src/platform_impl/linux/x11/util/input.rs

```diff
@@ -93,46 +94,64 @@ impl XConnection {
         device_id: c_int,
     ) -> Result<PointerState<'_>, XError> {
         unsafe {
-            let mut pointer_state: PointerState<'_> = mem::uninitialized();
-            pointer_state.xconn = self;
-            pointer_state.relative_to_window = (self.xinput2.XIQueryPointer)(
+            let mut root = MaybeUninit::uninit();
```

Comment:
I think it should be safe to zero-initialize the whole struct, then pass pointers to the individual members in it.

Reply:
`zeroed` is not safe for `PointerState`, as it contains a reference field, `&XConnection`.

I've instead zero-initialized each value before passing a reference to the FFI call.

Reply:
Afaik, if we set the `&XConnection` to self before calling `assume_init`, it should be safe.

It's the first example here: https://doc.rust-lang.org/nightly/std/mem/union.MaybeUninit.html#examples

Reply:
If `PointerState` contains a reference, `MaybeUninit` is probably the correct choice as long as the reference is filled with valid data before `assume_init()`. Apologies if my comments were misunderstood. The point I've been trying to make is that uninitialized data shouldn't be used as a blanket solution - it does have some valid cases. You just have to be extra careful - it's a footgun waiting to go off.

Reply:
But the field cannot be accessed without, at least, dereferencing a raw pointer to a mutable reference.

The documentation of `MaybeUninit` does not (to my knowledge) explicitly state that this would be undefined; however, the section on [field-by-field initialization](https://doc.rust-lang.org/nightly/std/mem/union.MaybeUninit.html#initializing-a-struct-field-by-field) states that initialization in this way is not possible. So, I take that to mean that it is undefined behavior to create a `&mut T` for uninitialized data.

Reply:
@murarth is correct, it's undefined behavior. I left an example below of how this specific case can be solved without any usage of uninitialized memory. 

> There is currently no supported way to create a raw pointer or reference to a field of a struct inside MaybeUninit<Struct>. That means it is not possible to create a struct by calling MaybeUninit::uninit::<Struct>() and then writing to its fields.